### ISSUE TITLE
Problems when organizing the alignment landmarks dataset.

### ISSUE THREAD

**Issue by oppo62258801** [CongWeilin/mtcnn-caffe#16]
hi, when I organize the alignment landmarks( 5 feature points on the face ) dataset, I use the CelebA dataset, I wonder if the information of list_bbox_celeba.txt/list_landmarks_align_celeba.txt/list_landmarks_celeba.txt should all be used to generate the pts.imdb file?

**Comment by CongWeilin** [comment]
you can should points on face to check out.

**Comment by oppo62258801** [comment]
thank you ,and could you tell me why the pNet(12net) and ONet(48net) use imdb data to feed into the layer,but the rNet(24net) use don't use the imd data？

**Comment by oppo62258801** [comment]
In file 24net/pythonLayer.py ,you encoding: label    = int(words[1]);  roi      = int(words[2]) both in line 60 and line 79. are you sure it's right? can you explain it simply?

**Comment by CongWeilin** [comment]
i use imdb to feed data into net.

**Comment by oppo62258801** [comment]
oh, I mean the code in your 24net/pythonLayer.py don't use the imdb.

**Comment by oppo62258801** [comment]
and I want to know if I want to use the imdb data, I should feed the imdb data into the net. Meanwhile I want to confirm that your pNet use the cls.imdb/roi.imdb/pts.imdb, rNet use the cls.imdb/roi.imdb and oNet use the cls.imdb/roi.imdb/pts.imdb? thank you!

**Comment by CongWeilin** [comment]
update pythonLayer.
sorry for forgetting to update them, thanks for reminding.
Q:and I want to know if I want to use the imdb data, I should feed the imdb data into the net? 
A:yes
Q:Meanwhile I want to confirm that your pNet use the cls.imdb/roi.imdb/pts.imdb, rNet use the cls.imdb/roi.imdb and oNet use the cls.imdb/roi.imdb/pts.imdb?
A: 12\24\48 all use cls+roi+pts

**Comment by oppo62258801** [comment]
THANK YOU！

**Comment by oppo62258801** [comment]
please help me! In your file "24net/train24.prototxt", you don't have anything about the landmarks loss. Does it mean the RNet(24Net) does't train the landmarks task?

**Comment by CongWeilin** [comment]
i have update 24net's pythonLayer.py. please re-pull

**Comment by oppo62258801** [comment]
thank you! I think your 24net/train24.prototxt should also be modified and does they should be very similar to the 12net/train12.prototxt and 48net/train48.prototxt?

**Comment by CongWeilin** [comment]
my mistake. i will update them as soon as possible. i might mass up the folder when gush to GitHub

**Comment by oppo62258801** [comment]
thank you! you are so kind. 
I almost train the dataset according to your method. but the train result is not so good. therefore, firstly I want to know if your label are the following forms:
12/negative/132458 -1
12/partive/125361 0 -0.08 -0.33 0.05 0.04
12/positive/363 1 0.06 0.04 0.22 -0.13
12/landmarks/26 0.31 0.35 0.47 0.36 0.34 0.45 0.32 0.49 0.48 0.49

**Comment by CongWeilin** [comment]
yes. the image you input should also normalize to (-1,1) to make gradient map more like a circle rather than ellipse. This training process needs some trick, because different loss will disturb each other. Good luck.

**Comment by oppo62258801** [comment]
I'm very puzzled at the codes in line 86 and line 104 in file prepare_data/gen_net_imdb.py, because:
in line 86, it is:  im -= 128;
in line 104,it is: im = (im - 127.5)/127.5.
I think they should be the same form?

**Comment by CongWeilin** [comment]
i change 127.5 into 128 can make it faster and 1/128 is quite close to 1/127.5.

**Comment by oppo62258801** [comment]
yes, but you may not get my point. I mean if you should chang line 86 into im = (im - 127.5)/127.5 like it in line 104, instead of im -= 128.
And I think only coding with im = (im - 127.5)/127.5 or im = (im - 128)/128(you think faster), can we normzlize an the input to (-1,1).

**Comment by CongWeilin** [comment]
i train the model with imdb. the imdb generate process use  im = (im - 127.5)/127.5.

**Comment by oppo62258801** [comment]
OK，I’m clear! then I should change it.

I have another problem. you set  im = (im - 127.5)/127.5 when generating imdb file, and in this case, the input image value is set to (-1,1). but in file pythonLayer.py,  you code  im -= self.mean (self.mean=128) within the class BatchLoader(object), is this meaningful?

**Comment by CongWeilin** [comment]
they are seperate part. if you use imdb, it will not read from image lists

**Comment by hzh8311** [comment]
Sorry to bother you, but where is the landmarks face preparation code?

**Comment by CongWeilin** [comment]
you can do this by yourself. downloads celeba and use the landmark list.................................................................................................................................................................................

**Comment by hzh8311** [comment]
Are you free for now, and QQ me at 389787216.

**Comment by oppo62258801** [comment]
you generate the imdb file using im = (im - 127.5)/127.5 and make the image to (-1,1), and when you read the image from the imdb file, the image you read is sure the value of (-1,1) ,so how can you code im -= self.mean(self.mean=128) ?

**Comment by oppo62258801** [comment]
of couse, I know they are seperate part,but the pythonLayer.py should read the imdb file from the imdb file generated by gen_net_imdb.py, are they?

**Comment by CongWeilin** [comment]
yes

**Comment by oppo62258801** [comment]
so I think it is contradictory for the above case? can you explain it for me?

**Comment by oppo62258801** [comment]
From your code, I think you did't use the val_set and only use the train_set in your training process. Then how to use the val_set through adding some corresponding code in the pythonlayer?
And I want to talk about something with you, so can you be a friend with me at weChat and my weChat name is my github name.

**Comment by CongWeilin** [comment]
to add cross_validate you need to modify `solver.prototxt`. you can search by yourself. i cut them off when upload

**Comment by oppo62258801** [comment]
ok,thank you！

**Comment by cuimiao187561** [comment]
Do you share your test set? @oppo62258801 
my email is 1906910382@qq.com
thank you very much

**Comment by wuzhiyang2016** [comment]
in gen_net_imdb.py ，images pre-process, some where ues im = (im - 127.5)/127.5  some where use im -= 128  ,,,why the processes are different?   which one is right?

**Comment by bpf476499866** [comment]
Hi, @oppo62258801 
..Sorry to bother you, but could you please tell me how did you generate the lables in your lable file "12/landmarks/26 0.31 0.35 0.47 0.36 0.34 0.45 0.32 0.49 0.48 0.49" ? how to calculate 0.31~0.49 whith the original face annotation (x,y,w,h) and landmark annotation (x1y1~x5y5) in celebA?

**Comment by bpf476499866** [comment]
Hi, @oppo62258801 
..Sorry to bother you, but could you please tell me how did you generate the lables in your lable file : 12/landmarks/26 0.31 0.35 0.47 0.36 0.34 0.45 0.32 0.49 0.48 0.49 ? how to calculate 0.31 to 0.49 whith the original face annotation (x,y,w,h) and landmark annotation (x1y1 to x5y5) in celebA?

**Comment by Zouhj** [comment]
@bpf476499866 hi, I am also confused, because my calculation result ranges from 0.30 to 0.79, so when I trained the model, something overflowed. I think maybe I have some errors when processing data of celeba. Do you solve this?

**Comment by bpf476499866** [comment]
@Zouhj 
no, not al all :(( 
I cut the face from celebleA, radomed the wide and  co-ordinates of the rect, and re-form the landmarks into relative co-ordinates to the randomed rect as "(x_landmark - x_rect_topleft) / width_rect";
but before the pts.imdb been used, other training problems stuck me……

#41 ←please have a glimpse of this, have you got any clue？

**Comment by Zouhj** [comment]
@bpf476499866 Thanks a lot! I looked at your problem, but I am very sorry I don't have any ideas.

**Comment by Zouhj** [comment]
@bpf476499866 and I trained the three tasks in 12net, however, the loss didn't decrease. and if the lr is 0.001~0.0001, loss will be nan. I changed lr very small, but I don't think something will be learnd by the small lr. It sames that congweiling choose a task to train each batch. I changed the code so that the three task can be traind each batch. and lr is 0.001, the net looks like can be converged. but I test the training model, result is very bad, too much rects(FP rects).

**Comment by bpf476499866** [comment]
@Zouhj 
!! unbelievable !! you made it !!
If you only tested p-net, then "many rects" is normal I think (at least... to some extent...), because the p-net is too tiny a net to get high Precision.
However what important is that the "true-positive face" rects must be included in those rects, thus the following  r-net and o-net will be able to pick them out....

Could you please share your "conjunct-task-batch" code if it's convenient to you? And I'm curious about your trained model, is there an image result sample?

Look forward to your reply, my personal mail is "476499866@qq.com" 
>-<

**Comment by wolf345** [comment]
@oppo62258801 12 / pts.imdb not found how did you fix it? Thank you

**Comment by bpf476499866** [comment]
@wolf345 
You need to generate it by yourself. 
python script is similar to widerface-imdb-generator
